src/distributions/weighted_index.rs

```diff
@@ -130,6 +130,72 @@ impl<X: SampleUniform + PartialOrd> WeightedIndex<X> {
         })
     }
 
+    /// Updates all weights by recalculating the index, without changing the number of weights.
+    ///
+    /// **NOTE:** if `weights` contains invalid elements (for example, `f64::NAN` in the case of
+    /// `X=f64`), the index might be left in a partially updated state. Sampling from this
+    /// partially updated index is undefined. It is the user's responsibility to not sample from
+    /// the index upon encountering an error. The index may be used again after assigning a new set 
+    /// of weights that do not result in an error.
+    pub fn assign_weights(&mut self, weights: &[X]) -> Result<(), WeightedError >
+    where
+        X: for<'a> ::core::ops::AddAssign<&'a X> + Clone + Default,
+    {
+        if weights.len() - 1 != self.cumulative_weights.len() {
```

Comment:
Prefer to add 1 on the RHS: it avoids panic-on-overflow in debug builds should `weights.len() == 0`.